pkg/ipam/metadata/manager.go

```diff
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package metadata
+
+import (
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/util/validation"
+
+	"github.com/cilium/cilium/pkg/annotation"
+	"github.com/cilium/cilium/pkg/hive"
+	ipamOption "github.com/cilium/cilium/pkg/ipam/option"
+	"github.com/cilium/cilium/pkg/k8s"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slim_core_v1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "ipam-metadata-manager")
+)
+
+type ManagerStoppedError struct{}
+
+func (m *ManagerStoppedError) Error() string {
+	return "ipam-metadata-manager has been stopped"
+}
+
+type ResourceNotFound struct {
+	Resource  string
+	Name      string
+	Namespace string
+}
+
+func (r *ResourceNotFound) Error() string {
+	name := r.Name
+	if r.Namespace != "" {
+		name = r.Namespace + "/" + r.Name
+	}
+	return fmt.Sprintf("Resource %s %q not found", r.Resource, name)
+}
+
+func (r *ResourceNotFound) Is(target error) bool {
+	targetErr, ok := target.(*ResourceNotFound)
+	if !ok {
+		return false
+	}
+	if r != nil && targetErr.Resource != "" {
+		return r.Resource == targetErr.Resource
+	}
+	return true
+}
+
+type Manager struct {
+	namespaceResource resource.Resource[*slim_core_v1.Namespace]
```

Comment:
Already asked but might be worth put in public :)
Since namespaceResource is not really used do we need to store it?

Reply:
Yes, we need `namespaceResource` (and it's `Pod` counterpart) to get access to the`namespaceStore`, which is what we actually want. Unfortunately, because obtaining the `Store` is a blocking operation (ref), we cannot obtain the store in the store in the `newIPAMMetadataManager` constructor and thus need to store the reference in the struct until `OnStart()` is called, at which point we are then allowed to obtain the `Store` reference.

My understanding of the hive framework is the daemon consumer of `ipamMetadata.Manager` should not observe a reference to it until it has been properly "started", so delaying the initialization of the stores into the `Start` hook is safe.